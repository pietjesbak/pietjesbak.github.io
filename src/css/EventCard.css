.facebook-card {
    position: relative;
}

.facebook-card .content {
    width: 100%;;
    max-width: 700px;
    display: inline-block;
}

.facebook-card .date {
    display: inline-block;
    background: #f7f7f7;
    padding: 10px;
    margin-bottom: 10px;
    margin-top: -10px;
    border-left: 3px solid #666;
}

.facebook-card .date.valid {
    border-left: 3px solid #103D5D;
}

.facebook-card .date.ongoing {
    border-left: 3px solid rgb(202, 139, 25);
}

.facebook-card .date.expired {
    border-left: 3px solid rgb(153, 25, 25);
}

.facebook-card .facebook {
    display: inline-block;
    font-size: 100px;
    vertical-align: top;
    color: #666;
}

.facebook-card .daZGjV {
    position: absolute;
    top: 10px;
    right: 10px;
}

.facebook-card .edit-event {
    background-color: #103D5D;
}

.facebook-card input[type="text"] {
    width: 100%;
    margin: 5px 0;
}

.facebook-card textarea {
    width: 100%;
    min-height: 200px;
    margin: 5px 0;
    resize: none;
}
