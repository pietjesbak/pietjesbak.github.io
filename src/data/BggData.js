/**
 * Little helper to filter out undefined elements.
 *
 * @param {*} element
 * @return {boolean} True if the element is not undefined.
 */
const filter = element => element !== undefined;

class BggGameData {
    /**
     * Data class that represents an item from boardgamegeek.
     *
     * @param {item} node An xml node from boardgamegeek.
     */
    constructor(node) {
        /**
         * This element's id.
         *
         * @type {number}
         */
        this.id = Number(node.attributes['objectid'].nodeValue);

        /**
         * This element's collection id.
         *
         * @type {number}
         */
        this.collectionId = Number(node.attributes['collid'].nodeValue);

        /**
         * The name of the game.
         *
         * @type {string}
         */
        this.name = node.getElementsByTagName('name')[0].childNodes[0].nodeValue;

        /**
         * The publish year.
         *
         * @type {number}
         */
        this.year = Number(node.getElementsByTagName('yearpublished')[0].childNodes[0].nodeValue);

        /**
         * Game image.
         *
         * @type {string}
         */
        this.image = node.getElementsByTagName('image')[0].childNodes[0].nodeValue;

        /**
         * Game thumbnail.
         *
         * @type {string}
         */
        this.thumbnail = node.getElementsByTagName('thumbnail')[0].childNodes[0].nodeValue;

        /**
         * Game stats.
         *
         * @type {BggStatsData}
         */
        this.stats = new BggStatsData(node.getElementsByTagName('stats')[0]);

        /**
         * Game details.
         *
         * @type {?BggDetailsData=}
         */
        this.details = undefined;

        /**
         * The amount of requests this month, not part of boardgame geek.
         * This should be filled in by firebase.
         *
         * @type {number}
         */
        this.requestsThisMonth = 0;

        /**
         * If this game was requested this month by the current use.
         * This should be filled in by firebase.
         */
        this.requestedByMe = false;
    }
}

class BggStatsData {
    /**
     * Data class that represents boardgamegeek stats.
     *
     * @param {item} node An xml node containing the stats.
     */
    constructor(node) {
        /**
         * Min players.
         *
         * @type {number}
         */
        this.minPlayers = Number(node.attributes['minplayers'].nodeValue);

        /**
         * Max players.
         *
         * @type {number}
         */
        this.maxPlayers = Number(node.attributes['maxplayers'].nodeValue);

        /**
         * Min play time.
         *
         * @type {number}
         */
        this.minPlaytime = Number(node.attributes['minplaytime'].nodeValue);

        /**
         * Max play time.
         *
         * @type {number}
         */
        this.maxPlaytime = Number(node.attributes['maxplaytime'].nodeValue);

        /**
         * Game rating.
         *
         * @type {number}
         */
        this.rating = Number(node.getElementsByTagName('average')[0].attributes['value'].nodeValue);
    }

    /**
     * Get players string.
     *
     * @returns {string}
     */
    get players() {
        if (this.minPlayers === this.maxPlayers) {
            return `${this.minPlayers} spelers`;
        }

        return `${this.minPlayers} - ${this.maxPlayers} spelers`;
    }

    /**
     * Get playtime string.
     *
     * @returns {string}
     */
    get playtime() {
        if (this.minPlaytime === this.maxPlaytime) {
            return `${this.minPlaytime} minuten`;
        }

        return `${this.minPlaytime} - ${this.maxPlaytime} minuten`;
    }
}

class BggDetailsData {
    /**
     * Data class that represents boardgamegeek details.
     *
     * @param {item}                      node  An xml node containing the details.
     * @param {Map.<number, BggGameData>} games All games.
     */
    constructor(node, games) {
        /**
         * Game description.
         *
         * @type {string}
         */
        this.description = node.getElementsByTagName('description')[0].childNodes[0].nodeValue;

        /**
         * Game categpries.
         *
         * @type {Array.<string>}
         */
        this.categories = [...node.getElementsByTagName('boardgamecategory')].filter(filter).map(element => element.childNodes[0].nodeValue);

        /**
         * Game family (theme).
         *
         * @type {string|undefined}
         */
        this.family = undefined;
        const familyNodes = node.getElementsByTagName('boardgamefamily');
        if (familyNodes.length === 1) {
            this.family = familyNodes[0].childNodes[0].nodeValue;
        }

        /**
         * Game mechanics.
         *
         * @type {Array.<string>}
         */
        this.mechanics = [...node.getElementsByTagName('boardgamemechanic')].filter(filter).map(element => element.childNodes[0].nodeValue);

        /**
         * Game expansions.
         *
         * @type {Map.<number, string>}
         */
        this.expansions = new Map();
        [...node.getElementsByTagName('boardgameexpansion')].filter(filter).forEach(element => this.expansions.set(Number(element.attributes['objectid'].nodeValue), element.childNodes[0].nodeValue));

        /**
         * Game domain (audience?).
         *
         * @type {Array.<string>}
         */
        this.domain = [...node.getElementsByTagName('boardgamesubdomain')].filter(filter).map(element => element.childNodes[0].nodeValue);

        /**
         * Owned expansions.
         *
         * @type {Map.<number, BggGameData>}
         */
        this.ownedExpansions = new Map();

        /**
         * Not owned expansions.
         *
         * @type {Map.<number, BggGameData>}
         */
        this.otherExpansions = new Map();

        this.updateOwnedExpansions(games);
    }

    /**
     * Updates the owned and other expansion maps.
     *
     * @param {Map.<number, BggGameData>} games All games.
     */
    updateOwnedExpansions(games) {
        const entries = [...this.expansions.entries()];

        this.ownedExpansions = new Map();
        entries
            .filter(id => games.has(id))
            .forEach(([id, name]) => this.ownedExpansions.set(id, name));

        this.otherExpansions = new Map();
        entries
            .filter(id => games.has(id) === false)
            .forEach(([id, name]) => this.otherExpansions.set(id, name));
    }

    /**
     * Get the description as an array of strings without html bits.
     *
     * @return {Array.<string>}
     */
    get descriptionArray() {
        const html = new DOMParser().parseFromString(this.description, 'text/html');

        let nextIsProbablyJunk = false;
        return [...html.body.childNodes].map(node => {
            // After a link there seems to be some truncated text (or a br element). Ignore that truncated text, a br returns an empty string anyways.
            if (nextIsProbablyJunk) {
                nextIsProbablyJunk = false;
                return '';
            }

            if (node.href !== undefined) {
                nextIsProbablyJunk = true;
                return node.href;
            }

            if (node.nodeValue !== null) {
                return node.nodeValue;
            }

            return '';
        });
    }
}

export {
    BggGameData,
    BggStatsData,
    BggDetailsData
};
