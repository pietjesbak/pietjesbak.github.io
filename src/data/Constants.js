/**
 * A heroku app running cors-anywhere to access some boardgame geek api endpoints.
 * https://github.com/Rob--W/cors-anywhere
 *
 * @const
 * @type {string}
 */
export const CORS_ANYWHERE_DYNO = 'https://pietjesbak.herokuapp.com/';

/**
 * The pietjesbak game collection on board game geek.
 *
 * @const
 * @type {string}
 */
export const PIETJESBAK_BBG_COLLECTION = 'https://www.boardgamegeek.com/xmlapi/collection/de%20pietjesbak?own=1';

/**
 * Board game geek game details api. Add the board game id. Requires the cors-anywhere dyno.
 *
 * @const
 * @type {string}
 */
export const BBG_GAME_API = 'https://www.boardgamegeek.com/xmlapi/boardgame/';

/**
 * Facebook pietjesbak events graph api.
 * Please don't steal my app secret k thanks.
 *
 * @const
 * @type {string}
 */
export const FACEBOOK_PIETJESBAK_EVENTS = 'https://graph.facebook.com/v2.10/gezelschapsspellenpietjesbak/events?access_token=874889636002007|788af4e5d543244dff7ee209987a2a57&format=json&limit=1';
