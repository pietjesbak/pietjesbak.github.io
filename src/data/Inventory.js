import { PIETJESBAK_BBG_COLLECTION } from './Constants';
import * as data from './BggData.js';
import firebase, { auth, provider } from '../Firebase.js';
import * as constants from './Constants.js';

class Inventory {
    constructor() {
        /**
         * A promise for all board game geek games.
         *
         * @private
         * @type {?Promise.<Map.<number, BggGameData>>}
         */
        this.fetchGamePromise_ = null;

        /**
         * A map of games you requested and their firebase key.
         *
         * @private
         * @type {?Map.<number, string>}
         */
        this.ownRequestedGames_ = null;

        /**
         * A map of all requested games and the amount of times they were requested this month.
         *
         * @private
         * @type {?Map.<number, number>}
         */
        this.requestedGames_ = null;

        /**
         * An array of callbacks for when the game lists update.
         *
         * @private
         * @type {Array.<function(Map.<number, BggGameData>)>}
         */
        this.changeListeners_ = [];

        /**
         * Your user object when logged in to firebase.
         *
         * @private
         * @type {?Object}
         */
        this.user_ = null;

        /**
         * All users of the pietjesbak. uid => name
         *
         * @private
         * @type {Map.<string, string>}
         */
        this.users_ = new Map();

        this.firebaseWaitForAuthChange_();
        this.getFirebaseRequests_();
        // this.getFirebaseUsers_();
    }

    /**
     * Get the board game collection from boardgame geek.
     *
     * @return {Promise.<Map.<number, BggGameDate>>}
     */
    async getGames() {
        if (this.fetchGamePromise_ === null) {
            this.fetchGamePromise_ = this.fetchGames_();
        }

        return await this.fetchGamePromise_;
    }

    /**
     * Toggles a request for a game. You need to be logged in for this.
     *
     * @param {*} game The game to toggle.
     * @return {Promise}
     */
    async toggleGame(game) {
        const date = new Date();
        const key = this.ownRequestedGames_.get(game.id);

        if (this.user_ === null) {
            throw new Error('You need to be logged in to request a game.');
        }

        if (key === undefined) {
            // Request it.
            await firebase.database().ref(`requests/${date.getFullYear()}/${date.getMonth()}/${this.user_.uid}/`).push(game.id);
        } else {
            // Remove.
            await firebase.database().ref(`requests/${date.getFullYear()}/${date.getMonth()}/${this.user_.uid}/${key}/`).remove();
        }
    }

    /**
     * Fetches all details for a game.
     *
     * @param {BggGameData} game The game to fetch details for.
     * @return {Promise}
     */
    async fetchGameDetails(game) {
        if (game.details !== undefined) {
            return;
        }

        game.details = null;

        try {
            let games = await this.getGames();
            let response = await fetch(constants.CORS_ANYWHERE_DYNO + constants.BBG_GAME_API + game.id);
            let xml = new DOMParser().parseFromString(await response.text(), 'text/xml');

            game.details = new data.BggDetailsData(xml.children[0].children[0], games);
        } catch (e) {
            game.details = undefined; // Reset
            throw e;
        }

        return game;
    }

    /**
     * Adds a listener that gets called everytime the games change. Contains: all game requests, game requests for the current user and all games.
     *
     * @param {function(Map.<number, BggGameData>)} listener The listener.
     * @return {Promise}
     */
    async addChangeListener(listener) {
        this.changeListeners_.push(listener);

        if (this.requestedGames_ !== null) {
            listener.call(this, await this.getGames());
        }
    }

    /**
     * Removes a listener that was added using addChangeListener.
     *
     * @param {function(Map.<number, BggGameData>)} listener The previously added listener.
     */
    removeChangeListener(listener) {
        this.changeListeners_.splice(this.changeListeners_.indexOf(listener), 1);
    }

    /**
     * Get the currently logged in user.
     *
     * @return {?Object}
     */
    get user() {
        return this.user_;
    }

    /**
     * Get all users.
     *
     * @return {Map.<number, Object>}
     */
    get users() {
        return this.users_;
    }

    /***************************************************************
     * AUTHENTICATION                                              *
     ***************************************************************/

    /**
     * Logs out from firebase.
     *
     * @returns {Promise}
     */
    logout() {
        return new Promise(resolve => {
            auth.signOut().then(() => {
                this.user_ = null;
                resolve();
            });
        })
    }

    /**
     * Logs in to firebase.
     *
     * @return {Promise.<Object>}
     */
    login() {
        return new Promise((resolve, reject) => {
            auth.signInWithRedirect(provider).then(() => {
                auth.getRedirectResult().then(result => {
                    this.user_ = result.user;

                    resolve(this.user_);
                }).catch(e => {
                    console.error(e);
                    reject();
                });
            });
        });
    }

    /***************************************************************
     * PRIVATE STUFF                                               *
     ***************************************************************/

    /**
    * Wraps window.setTimeout into a promise.
    *
    * @private
    * @param {number} ms ms to sleep.
    */
    async sleep_(ms) {
        return new Promise(resolve => {
            window.setTimeout(resolve, ms);
        });
    }

    /**
     * Fetches the board game collection from boardgame geek.
     *
     * @private
     * @return {Promise.<Map.<number, BbgGameDate>>}
     */
    async fetchGames_() {
        let response;
        do {
            response = await fetch(PIETJESBAK_BBG_COLLECTION);
            if (response.status !== 200) {
                await this.sleep_(2000);
            }
        } while (response.status !== 200);

        let xml = new DOMParser().parseFromString(await response.text(), 'text/xml');

        let games = new Map();
        for (let child of xml.children[0].children) {
            let game = new data.BggGameData(child);
            games.set(game.id, game);
        }

        return games;
    }

    /**
     * Updates the user when the auth state changes.
     *
     * @private
     */
    firebaseWaitForAuthChange_() {
        auth.onAuthStateChanged(user => {
            if (user) {
                this.user_ = user;
            } else {
                this.user_ = null;
            }
        });
    }

    /**
     * Get all game requests from firebase.
     *
     * @private
     */
    async getFirebaseRequests_() {
        const date = new Date();
        firebase.database().ref(`requests/${date.getFullYear()}/${date.getMonth()}/`).on('value', async snapshot => {
            const value = snapshot.val();
            const ownUid = this.user !== null ? this.user.uid : undefined;
            this.ownRequestedGames_ = new Map();
            this.requestedGames_ = new Map();

            for (const uid in value) {
                const requests = [...new Set(Object.values(value[uid]))];
                if (uid === ownUid) {
                    const keys = Object.keys(value[uid]);
                    for (let i = 0; i < requests.length; i++) {
                        this.ownRequestedGames_.set(requests[i], keys[i]);
                    }
                }

                for (let i = 0; i < requests.length; i++) {
                    this.requestedGames_.set(requests[i], 1 + (this.requestedGames_.get(requests[i]) || 0));
                }
            }

            let games = await this.getGames();
            for (let [k, v] of games.entries()) {
                v.requestsThisMonth = this.requestedGames_.get(k) || 0;
                v.requestedByMe = this.ownRequestedGames_.has(k);
            }

            this.changeListeners_.forEach(listener => listener.call(this, games));
        });
    }

    /**
     * Get all users from firebase.
     *
     * @private
     */
    getFirebaseUsers_() {
        firebase.database().ref(`users/`).once('value', snapshot => {
            const value = snapshot.val();
            this.users_ = new Map();
            for (let uid in value) {
                this.users_.set(uid, value[uid]);
            }

            // You only have access to the users after logging in so the timing should always be right.
            if (this.user !== null && this.users_.has(this.user.uid) === false) {
                firebase.database().ref(`users/${this.user.uid}/`).set(this.user.displayName);
            }
        });
    }
}

export default new Inventory();
