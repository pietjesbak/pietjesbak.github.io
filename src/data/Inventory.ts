import { auth, database, provider } from '../Firebase';
import * as data from './BggData';
import { PIETJESBAK_BBG_COLLECTION } from './Constants';
import * as constants from './Constants';

export enum ChangeEvent {
    GAME_DATA,
    USER
}

export class Inventory {
    /**
     * A promise for all board game geek games.
     */
    private fetchGamePromise_: Promise<Map<number, data.BggGameData>> | null = null;

    /**
     * A map of games you requested and their firebase key.
     */
    private ownRequestedGames_: Map<number, string> | null = null;

    /**
     * A map of all requested games and the amount of times they were requested this month.
     */
    private requestedGames_: Map<number, number> | null = null;

    /**
     * An array of callbacks for when the game lists update.
     */
    private changeListeners_: Map<ChangeEvent, Array<(data: any) => void>> = new Map();

    /**
     * Your user object when logged in to firebase.
     */
    private user_: FirebaseUser | null = null;

    /**
     * All users of the pietjesbak. uid => name
     */
    private users_: Map<string, string> = new Map();

    /**
     * The cached firebase message.
     */
    private message_: FirebaseMessage | null = null;

    constructor() {
        this.firebaseWaitForAuthChange_();
        this.getFirebaseRequests_();
        // this.getFirebaseUsers_();
    }

    /**
     * Get the board game collection from boardgame geek.
     */
    async getGames() {
        if (this.fetchGamePromise_ === null) {
            this.fetchGamePromise_ = this.fetchGames_();
        }

        return await this.fetchGamePromise_;
    }

    /**
     * Toggles a request for a game. You need to be logged in for this.
     *
     *  @param game The game to toggle.
     */
    async toggleGame(game: data.BggGameData) {
        const date = (await this.getNextEventDate()).date;
        const key = this.ownRequestedGames_!.get(game.id);

        if (this.user_ === null) {
            throw new Error('You need to be logged in to request a game.');
        }

        if (key === undefined) {
            // Request it.
            await database.ref(`requests/${date.getFullYear()}/${date.getMonth()}/${this.user_.uid}/`).push(game.id);
        } else {
            // Remove.
            await database.ref(`requests/${date.getFullYear()}/${date.getMonth()}/${this.user_.uid}/${key}/`).remove();
        }
    }

    /**
     * Fetches all details for a game.
     *
     * @param game The game to fetch details for.
     */
    async fetchGameDetails(game: data.BggGameData) {
        if (game.details !== undefined) {
            return;
        }

        game.details = null;

        try {
            const games = await this.getGames();
            const response = await fetch(constants.CORS_ANYWHERE_DYNO + constants.BBG_GAME_API + game.id);
            const xml = new DOMParser().parseFromString(await response.text(), 'text/xml');

            game.details = new data.BggDetailsData(xml.children[0].children[0], games);
        } catch (e) {
            game.details = undefined; // Reset
            throw e;
        }

        return game;
    }

    /**
     * Adds a listener that gets called everytime the games change. Contains: all game requests, game requests for the current user and all games.
     *
     * @param event    The type of event.
     * @param listener The listener.
     */
    async addChangeListener(event: ChangeEvent, listener: (data: any) => void) {
        if (this.changeListeners_.has(event) === false) {
            this.changeListeners_.set(event, [listener]);
        } else {
            this.changeListeners_.get(event)!.push(listener);
        }

        if (event === ChangeEvent.GAME_DATA) {
            if (this.requestedGames_ !== null) {
                listener.call(this, await this.getGames());
            }
        }
    }

    /**
     * Removes a listener that was added using addChangeListener.
     *
     * @param event    The type of event.
     * @param listener The previously added listener.
     */
    removeChangeListener(event: ChangeEvent, listener: (data: any) => void) {
        if (this.changeListeners_.has(event) === true) {
            this.changeListeners_.get(event)!.splice(this.changeListeners_.get(event)!.indexOf(listener), 1);
        }
    }

    /**
     * Get the currently logged in user.
     */
    get user() {
        return this.user_;
    }

    /**
     * Get all users.
     */
    get users() {
        return this.users_;
    }

    /**
     * Get a promise containing the next event date and if this date has been confirmed.
     */
    async getNextEventDate() {
        let confirmed = true;
        let startTime;

        try {
            const event = await this.getMessage();
            startTime = new Date(event.date);
        } catch (r) {
            console.error('Failed to get next event date.');
            startTime = new Date();
        }

        const thresholdDate = new Date();
        thresholdDate.setDate(thresholdDate.getDate() + 1);

        // Adjust the start time if the event has passed.
        for (let i = 0; i < 2 && thresholdDate > startTime; i++) {
            confirmed = false;
            startTime.setMonth(thresholdDate.getMonth() + i); // Check this month and the next month for the first available event date.
            startTime.setDate(1); // Set to first day of this month.
            startTime.setDate(2 * 7 + (5 - startTime.getDay() + 7) % 7 + 1); // Figure out when the 3rd friday is from here.
        }

        return {
            confirmed,
            date: startTime
        };
    }

    /**
     * Get the message from firebase.
     */
    async getMessage(): Promise<FirebaseMessage> {
        if (this.message_ === null) {
            const message = await database.ref('message').once('value');
            const val = message.val();
            if (val === null) {
                throw new Error('Message is empty!');
            }

            this.message_ = {
                title: val.title,
                body: val.body,
                date: new Date(val.date)
            }
        }

        return this.message_;
    }

    /**
     * Update the event in firebase.
     *
     * @param title The event title.
     * @param body The event body.
     * @param date The event date.
     */
    async updateMessage(title: string, body: string, date: number) {
        await database.ref('message').set({
            title,
            body,
            date
        });

        this.message_ = null;
    }

    /***************************************************************
     * AUTHENTICATION                                              *
     ***************************************************************/

    /**
     * Logs out from firebase.
     */
    logout() {
        return new Promise(resolve => {
            auth.signOut().then(() => {
                this.user_ = null;
                resolve();
            });
        })
    }

    /**
     * Logs in to firebase.
     */
    login(): Promise<FirebaseUser> {
        return new Promise((resolve, reject) => {
            auth.signInWithRedirect(provider).then(() => {
                auth.getRedirectResult().then(result => {
                    this.user_ = result.user;

                    resolve(this.user_!);
                }).catch(e => {
                    console.error(e);
                    reject();
                });
            });
        });
    }

    /***************************************************************
     * PRIVATE STUFF                                               *
     ***************************************************************/

    /**
     * Wraps window.setTimeout into a promise.
     *
     * @param ms ms to sleep.
     */
    private async sleep_(ms: number) {
        return new Promise(resolve => {
            window.setTimeout(resolve, ms);
        });
    }

    /**
     * Fetches the board game collection from boardgame geek.
     */
    private async fetchGames_() {
        let response;
        do {
            try {
                response = await fetch(PIETJESBAK_BBG_COLLECTION);
                if (response.status !== 200) {
                    throw new Error('Collection not yet available.');
                }
            } catch (e) {
                console.log(`Trying again in 3 seconds: ${e}`);
                await this.sleep_(3000);
            }
        } while (response === undefined || response.status !== 200);

        const xml = new DOMParser().parseFromString(await response.text(), 'text/xml');

        const games = new Map();
        for (const child of xml.children[0].children as any) {
            const game = new data.BggGameData(child);
            games.set(game.id, game);
        }

        return games;
    }

    /**
     * Updates the user when the auth state changes.
     */
    private firebaseWaitForAuthChange_() {
        auth.onAuthStateChanged(async user => {
            if (user) {
                this.user_ = user as FirebaseUser;

                const admins = await database.ref('admins').once('value');
                this.user_.admin = admins.val()[user.uid] !== undefined;
            } else {
                this.user_ = null;
            }

            if (this.changeListeners_.has(ChangeEvent.USER) === true) {
                this.changeListeners_.get(ChangeEvent.USER)!.forEach(listener => listener.call(this, this.user_));
            }
        });
    }

    /**
     * Get all game requests from firebase.
     */
    private async getFirebaseRequests_() {
        const date = (await this.getNextEventDate()).date;
        database.ref(`requests/${date.getFullYear()}/${date.getMonth()}/`).on('value', async snapshot => {
            if (snapshot !== null) {
                const value = snapshot.val() as { [key: string]: number };
                const ownUid = this.user_ !== null ? this.user_.uid : undefined;
                this.ownRequestedGames_ = new Map();
                this.requestedGames_ = new Map();

                for (const uid in value) {
                    const requests = [...new Set(Object.values(value[uid]))];
                    if (uid === ownUid) {
                        const keys = Object.keys(value[uid]);
                        for (let i = 0; i < requests.length; i++) {
                            this.ownRequestedGames_.set(requests[i], keys[i]);
                        }
                    }

                    for (let i = 0; i < requests.length; i++) {
                        this.requestedGames_.set(requests[i], 1 + (this.requestedGames_.get(requests[i]) || 0));
                    }
                }

                const games = await this.getGames();
                for (const [k, v] of games.entries()) {
                    v.requestsThisMonth = this.requestedGames_.get(k) || 0;
                    v.requestedByMe = this.ownRequestedGames_.has(k);
                }

                if (this.changeListeners_.has(ChangeEvent.GAME_DATA) === true) {
                    this.changeListeners_.get(ChangeEvent.GAME_DATA)!.forEach(listener => listener.call(this, games));
                }
            }
        });
    }

    /**
     * Get all users from firebase.
     */
    // private getFirebaseUsers_() {
    //     database.ref(`users/`).once('value', snapshot => {
    //         const value = snapshot.val();
    //         this.users_ = new Map();
    //         for (const uid in value) {
    //             this.users_.set(uid, value[uid]);
    //         }

    //         // You only have access to the users after logging in so the timing should always be right.
    //         if (this.user !== null && this.users_.has(this.user.uid) === false) {
    //             database.ref(`users/${this.user.uid}/`).set(this.user.displayName);
    //         }
    //     });
    // }
}
