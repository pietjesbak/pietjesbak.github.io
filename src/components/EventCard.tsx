import "../css/EventCard.css";

import classNames from "classnames";
import * as React from "react";
import { Inventory } from "../data/Inventory";
import { readableDate } from "../Helpers";
import { withInventory } from "../InventoryProvider";
import { Container } from "./Container";
import IconButton, { IconButtonBehavour } from "./IconButton";
import { TextPlaceholder } from "./TextPlaceholder";

export interface State {
  editing: boolean;
  event: FirebaseMessage | null;
  error: boolean;
}

export interface Props {
  inventory: Inventory;
}

export class EventCard extends React.Component<
  Props & React.ClassAttributes<EventCard>,
  State
> {
  constructor(props: Props & React.ClassAttributes<EventCard>) {
    super(props);
    this.state = {
      editing: false,
      event: null,
      error: false,
    };
  }

  componentDidMount() {
    this.updateMessage();
  }

  async updateMessage() {
    try {
      const event = await this.props.inventory.getMessage();
      this.setState({ event });
    } catch (e) {
      console.error(e);
      this.setState({ error: true });
    }
  }

  toggleMessage = async () => {
    if (this.state.editing === true) {
      await this.props.inventory.updateMessage(
        (this.refs.title as HTMLInputElement).value,
        (this.refs.body as HTMLTextAreaElement).value,
        new Date((this.refs.date as HTMLInputElement).value).getTime()
      );

      this.updateMessage();
    }

    this.setState({ editing: this.state.editing === false });
  };

  renderDate(date: Date) {
    const readable = readableDate(date, true);

    if (Date.now() - date.getTime() < 0) {
      return (
        <div className="date valid">
          <span>Op {readable}</span>
        </div>
      );
    }

    if (Date.now() - date.getTime() < 1000 * 60 * 60 * 5) {
      return (
        <div className="date ongoing">
          <span>Op {readable}</span>
          <br />
          <span>
            <strong>Nu aan de gang!</strong>
          </span>
        </div>
      );
    }

    return (
      <div className="date expired">
        <span>Op {readable}</span>
        <br />
        <span>
          Dit event is voorbij, nieuwe events worden meestal 2 weken op voorhand
          gepland.
        </span>
      </div>
    );
  }

  renderEvent(event: FirebaseMessage | null, error: boolean) {
    if (error === true) {
      return (
        <div className="event">
          <h2>Er lijkt geen volgend event gepland te zijn.</h2>
        </div>
      );
    }

    if (event === null) {
      return (
        <div className="event">
          <TextPlaceholder paragraphs={2} paragraphSize={3} />
        </div>
      );
    }

    if (this.state.editing === false) {
      return (
        <div className="event">
          <h2>{event.title}</h2>
          {this.renderDate(event.date)}
          <br />

          {event.body.split("\n").map((item, key) => {
            return (
              <span key={key}>
                {item}
                <br />
              </span>
            );
          })}
        </div>
      );
    } else {
      return this.renderEditableEvent(event);
    }
  }

  renderEditableEvent(event: FirebaseMessage) {
    return (
      <div className="content">
        <input type="text" defaultValue={event.title} ref="title" />
        <input
          type="datetime-local"
          defaultValue={new Date(
            event.date.getTime() - event.date.getTimezoneOffset() * 60000
          )
            .toISOString()
            .slice(0, -1)}
          ref="date"
        />
        <textarea defaultValue={event.body} ref="body" />
      </div>
    );
  }

  render() {
    return (
      <Container
        className={classNames("event-card", { error: this.state.error })}
      >
        <a
          className="facebook"
          href="https://www.facebook.com/gezelschapsspellenpietjesbak/"
        >
          <i className="icon-facebook-squared" />
        </a>

        {this.renderEvent(this.state.event, this.state.error)}
        {this.props.inventory.user !== null &&
        this.props.inventory.user.admin === true ? (
          <IconButton
            subClass="edit-event"
            action={this.toggleMessage}
            icon="cog"
            text={this.state.editing ? "Opslaan" : "Aanpassen"}
            behaviour={IconButtonBehavour.SMALL}
          />
        ) : (
          <span />
        )}
      </Container>
    );
  }
}

export default withInventory(EventCard);
