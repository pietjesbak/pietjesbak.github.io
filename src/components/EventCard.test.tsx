import * as enzyme from 'enzyme';
import * as React from 'react';

import { Inventory } from '../data/Inventory';
import { EventCard } from './EventCard';

function getMockInventory(date: Date = new Date()) {
    const inventory = new Inventory();
    inventory.getMessage = () => {
        return Promise.resolve({
            title: 'Test event',
            body: 'The new test event starts now!',
            date
        });
    }

    return inventory;
}

describe('<EventCard />', () => {
    it('Renders a placeholder', () => {
        const inventory = getMockInventory();
        const card = enzyme.shallow(<EventCard inventory={inventory} />);

        expect(card).toMatchSnapshot();
    });

    it('Renders an event', async () => {
        const inventory = getMockInventory();
        const card = enzyme.shallow(<EventCard inventory={inventory} />);

        await Promise.resolve();
        expect(card).toMatchSnapshot();
    });

    it('Renders an ongoing event', async () => {
        const inventory = getMockInventory(new Date('September 19 2018 9:30'));
        const card = enzyme.shallow(<EventCard inventory={inventory} />);

        await Promise.resolve();
        expect(card.find('.date')).toMatchSnapshot();
    });

    it('Renders an passed event', async () => {
        const inventory = getMockInventory(new Date('September 18 2018 11:00'));
        const card = enzyme.shallow(<EventCard inventory={inventory} />);

        await Promise.resolve();
        expect(card.find('.date')).toMatchSnapshot();
    });

    it('Renders an event that has yet to come', async () => {
        const inventory = getMockInventory(new Date('September 20 2018 11:00'));
        const card = enzyme.shallow(<EventCard inventory={inventory} />);

        await Promise.resolve();
        expect(card.find('.date')).toMatchSnapshot();
    });
});
