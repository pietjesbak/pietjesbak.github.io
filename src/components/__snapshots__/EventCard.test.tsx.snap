// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EventCard /> Renders a placeholder 1`] = `
<Container
  className="event-card"
>
  <a
    className="facebook"
    href="https://www.facebook.com/gezelschapsspellenpietjesbak/"
  >
    <i
      className="icon-facebook-squared"
    />
  </a>
  <div
    className="event"
  >
    <TextPlaceholder
      paragraphSize={3}
      paragraphs={2}
    />
  </div>
  <span />
</Container>
`;

exports[`<EventCard /> Renders an event 1`] = `
<Container
  className="event-card"
>
  <a
    className="facebook"
    href="https://www.facebook.com/gezelschapsspellenpietjesbak/"
  >
    <i
      className="icon-facebook-squared"
    />
  </a>
  <div
    className="event"
  >
    <h2>
      Test event
    </h2>
    <div
      className="date ongoing"
    >
      <span>
        Op 
        woensdag 19 september 2018 10:20
      </span>
      <br />
      <span>
        <strong>
          Nu aan de gang!
        </strong>
      </span>
    </div>
    <br />
    <span
      key="0"
    >
      The new test event starts now!
      <br />
    </span>
  </div>
  <span />
</Container>
`;

exports[`<EventCard /> Renders an event that has yet to come 1`] = `
<div
  className="date valid"
>
  <span>
    Op 
    donderdag 20 september 2018 11:00
  </span>
</div>
`;

exports[`<EventCard /> Renders an ongoing event 1`] = `
<div
  className="date ongoing"
>
  <span>
    Op 
    woensdag 19 september 2018 09:30
  </span>
  <br />
  <span>
    <strong>
      Nu aan de gang!
    </strong>
  </span>
</div>
`;

exports[`<EventCard /> Renders an passed event 1`] = `
<div
  className="date expired"
>
  <span>
    Op 
    dinsdag 18 september 2018 11:00
  </span>
  <br />
  <span>
    Dit event is voorbij, nieuwe events worden meestal 2 weken op voorhand gepland.
  </span>
</div>
`;
